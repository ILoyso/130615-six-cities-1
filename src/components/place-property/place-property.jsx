import React from 'react';
import PropTypes from 'prop-types';
import {connect} from 'react-redux';

import history from '../../utils/history';
import {Routes} from '../../constants/routes';
import ReviewsList from '../reviews-list/reviews-list.jsx';
import PlacesList from '../places-list/places-list.jsx';
import PlacesMap from '../places-map/places-map.jsx';

import {getRatingInPercent} from '../../utils/utils';
import {Operation} from '../../reducer/data/data';
import {getComments, getNearestPlaces} from '../../reducer/data/selectors';
import withActiveItem from '../../hocs/with-active-item/with-active-item';
import {getAuthorizationStatus} from '../../reducer/user/selectors';


const PlacesListWrapped = withActiveItem(PlacesList);


// PlaceProperty component
class PlaceProperty extends React.PureComponent {

  /**
   * Method for render component
   * @return {*}
   */
  render() {
    const {
      comments,
      isAuthorizationRequired,
      nearestPlaces,
      place,
      setFavorite
    } = this.props;

    const {
      bedrooms,
      city,
      description,
      goods,
      host,
      id,
      images,
      isFavorite,
      isPremium,
      maxAdults,
      price,
      rating,
      title,
      type,
    } = place;

    return <main className="page__main page__main--property">
      <section className="property">
        <div className="property__gallery-container container">
          <div className="property__gallery">
            {images.map((image, index) => <div className="property__image-wrapper" key={index}>
              <img className="property__image" src={image} alt="Photo studio" />
            </div>)}
          </div>
        </div>
        <div className="property__container container">
          <div className="property__wrapper">
            {isPremium && <div className="property__mark">
              <span>Premium</span>
            </div>}
            <div className="property__name-wrapper">
              <h1 className="property__name">{title}</h1>
              <button
                className={`property__bookmark-button ${isFavorite && `property__bookmark-button--active`} button`}
                onClick={() => {
                  if (isAuthorizationRequired) {
                    history.push(Routes.LOGIN);
                  } else {
                    setFavorite(place);
                  }
                }}
                type="button"
              >
                <svg className="property__bookmark-icon" width="31" height="33">
                  <use xlinkHref="#icon-bookmark"></use>
                </svg>
                <span className="visually-hidden">To bookmarks</span>
              </button>
            </div>
            <div className="property__rating rating">
              <div className="property__stars rating__stars">
                <span style={{width: `${getRatingInPercent(rating)}%`}}></span>
                <span className="visually-hidden">Rating</span>
              </div>
              <span className="property__rating-value rating__value">{Math.round(rating)}</span>
            </div>
            <ul className="property__features">
              <li className="property__feature property__feature--entire">
                {type}
              </li>
              <li className="property__feature property__feature--bedrooms">
                {bedrooms} Bedrooms
              </li>
              <li className="property__feature property__feature--adults">
                Max {maxAdults} adults
              </li>
            </ul>
            <div className="property__price">
              <b className="property__price-value">&euro;{price}</b>
              <span className="property__price-text">&nbsp;night</span>
            </div>
            <div className="property__inside">
              <h2 className="property__inside-title">What&apos;s inside</h2>
              <ul className="property__inside-list">
                {goods.map((good, index) => <li className="property__inside-item" key={index}>
                  {good}
                </li>)}
              </ul>
            </div>
            <div className="property__host">
              <h2 className="property__host-title">Meet the host</h2>
              <div className="property__host-user user">
                <div className="property__avatar-wrapper property__avatar-wrapper--pro user__avatar-wrapper">
                  <img className="property__avatar user__avatar" src={host.avatar} width="74" height="74" alt="Host avatar" />
                </div>
                <span className="property__user-name">
                  {host.name}
                </span>
                {host.isPro && <span className="property__user-status">Pro</span>}
              </div>
              <div className="property__description">
                <p className="property__text">
                  {description}
                </p>
              </div>
            </div>

            <ReviewsList
              comments={comments}
              id={id}
              isAuthorizationRequired={isAuthorizationRequired}
            />
          </div>
        </div>

        <PlacesMap
          city={city}
          className={`property`}
          places={nearestPlaces}
        />
      </section>
      <div className="container">
        <section className="near-places places">
          <h2 className="near-places__title">Other places in the neighbourhood</h2>

          <PlacesListWrapped
            classHelper={`near`}
            places={nearestPlaces}
          />
        </section>
      </div>
    </main>;
  }

  /**
   * Method is invoked immediately after a component is mounted (inserted into the tree)
   * Here is comments downloaded
   */
  componentDidMount() {
    const {place} = this.props;
    this.props.loadComments(place.id);
  }
}


PlaceProperty.propTypes = {
  comments: PropTypes.arrayOf(PropTypes.shape({
    user: PropTypes.shape({
      isPro: PropTypes.bool.isRequired,
      name: PropTypes.string.isRequired,
      avatar: PropTypes.string.isRequired,
    }).isRequired,
    rating: PropTypes.number.isRequired,
    comment: PropTypes.string.isRequired,
    date: PropTypes.string.isRequired,
  })).isRequired,
  isAuthorizationRequired: PropTypes.bool.isRequired,
  loadComments: PropTypes.func.isRequired,
  nearestPlaces: PropTypes.arrayOf(PropTypes.shape({
    bedrooms: PropTypes.number.isRequired,
    coordinates: PropTypes.arrayOf(PropTypes.number).isRequired,
    description: PropTypes.string.isRequired,
    goods: PropTypes.arrayOf(PropTypes.string).isRequired,
    host: PropTypes.shape({
      avatar: PropTypes.string.isRequired,
      isPro: PropTypes.bool.isRequired,
      name: PropTypes.string.isRequired,
    }).isRequired,
    id: PropTypes.number.isRequired,
    img: PropTypes.string.isRequired,
    images: PropTypes.arrayOf(PropTypes.string).isRequired,
    isPremium: PropTypes.bool.isRequired,
    maxAdults: PropTypes.number.isRequired,
    price: PropTypes.number.isRequired,
    rating: PropTypes.number.isRequired,
    title: PropTypes.string.isRequired,
    type: PropTypes.string.isRequired,
  })).isRequired,
  place: PropTypes.shape({
    bedrooms: PropTypes.number.isRequired,
    city: PropTypes.string.isRequired,
    description: PropTypes.string.isRequired,
    goods: PropTypes.arrayOf(PropTypes.string).isRequired,
    host: PropTypes.shape({
      avatar: PropTypes.string.isRequired,
      isPro: PropTypes.bool.isRequired,
      name: PropTypes.string.isRequired,
    }).isRequired,
    id: PropTypes.number.isRequired,
    images: PropTypes.arrayOf(PropTypes.string).isRequired,
    isFavorite: PropTypes.bool.isRequired,
    isPremium: PropTypes.bool.isRequired,
    maxAdults: PropTypes.number.isRequired,
    price: PropTypes.number.isRequired,
    rating: PropTypes.number.isRequired,
    title: PropTypes.string.isRequired,
    type: PropTypes.string.isRequired,
  }).isRequired,
  setFavorite: PropTypes.func.isRequired,
};


/**
 * Function for connect state with current component
 * @param {Object} state
 * @param {Object} ownProps
 * @return {Object}
 */
const mapStateToProps = (state, ownProps) => Object.assign({}, ownProps, {
  comments: getComments(state),
  isAuthorizationRequired: getAuthorizationStatus(state),
  nearestPlaces: getNearestPlaces(state, ownProps.place.id)
});


/**
 * Function for connect action creator methods with current component
 * @param {Function} dispatch
 * @return {Object}
 */
const mapDispatchToProps = (dispatch) => ({
  loadComments: (id) => dispatch(Operation.loadComments(id)),
  setFavorite: (place) => dispatch(Operation.setFavorite(place))
});


export {PlaceProperty};

export default connect(mapStateToProps, mapDispatchToProps)(PlaceProperty);
