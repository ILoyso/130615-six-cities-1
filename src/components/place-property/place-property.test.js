import React from 'react';
import renderer from 'react-test-renderer';
import {MemoryRouter} from 'react-router-dom';

import {PlaceProperty} from './place-property.jsx';


const placeMock = {
  bedrooms: 5,
  city: `Test City`,
  description: `Test Description`,
  goods: [`Good1`, `Good2`],
  host: {
    avatar: `testAvatar`,
    isPro: true,
    name: `testName`,
  },
  id: 1,
  images: [`Img1`, `Img2`],
  isPremium: true,
  maxAdults: 5,
  price: 200,
  rating: 3.1,
  title: `Test Title`,
  type: `Test Type`,
};

const nearestPlaceMock = [{
  bedrooms: 5,
  coordinates: [100, 300],
  description: `Test Description`,
  goods: [`Good1`, `Good2`],
  host: {
    avatar: `testAvatar`,
    isPro: true,
    name: `testName`,
  },
  id: 1,
  img: `test img`,
  images: [`Img1`, `Img2`],
  isPremium: true,
  maxAdults: 5,
  price: 200,
  rating: 3.1,
  title: `Test Title`,
  type: `Test Type`,
}];

const commentMock = [{
  user: {
    isPro: true,
    name: `name`,
    avatar: `avatar`,
  },
  rating: 4,
  comment: `comment`,
  date: `date`,
}];

it(`PlaceProperty correctly renders`, () => {
  const placeProperty = renderer
    .create(<MemoryRouter>
      <PlaceProperty
        comments={commentMock}
        isAuthorizationRequired={true}
        loadComments={jest.fn()}
        nearestPlaces={nearestPlaceMock}
        place={placeMock}
      />
    </MemoryRouter>)
    .toJSON();

  expect(placeProperty).toMatchSnapshot();
});
