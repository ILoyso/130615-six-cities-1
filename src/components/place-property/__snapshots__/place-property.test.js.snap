// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PlaceProperty Component PlaceProperty correctly renders 1`] = `
<main
  className="page__main page__main--property"
>
  <section
    className="property"
  >
    <div
      className="property__gallery-container container"
    >
      <div
        className="property__gallery"
      >
        <div
          className="property__image-wrapper"
          key="0"
        >
          <img
            alt="Photo studio"
            className="property__image"
            src="Img1"
          />
        </div>
        <div
          className="property__image-wrapper"
          key="1"
        >
          <img
            alt="Photo studio"
            className="property__image"
            src="Img2"
          />
        </div>
      </div>
    </div>
    <div
      className="property__container container"
    >
      <div
        className="property__wrapper"
      >
        <div
          className="property__mark"
        >
          <span>
            Premium
          </span>
        </div>
        <div
          className="property__name-wrapper"
        >
          <h1
            className="property__name"
          >
            Test Title
          </h1>
          <button
            className="property__bookmark-button property__bookmark-button--active button"
            onClick={[Function]}
            type="button"
          >
            <svg
              className="property__bookmark-icon"
              height="33"
              width="31"
            >
              <use
                xlinkHref="#icon-bookmark"
              />
            </svg>
            <span
              className="visually-hidden"
            >
              To bookmarks
            </span>
          </button>
        </div>
        <div
          className="property__rating rating"
        >
          <div
            className="property__stars rating__stars"
          >
            <span
              style={
                Object {
                  "width": "60%",
                }
              }
            />
            <span
              className="visually-hidden"
            >
              Rating
            </span>
          </div>
          <span
            className="property__rating-value rating__value"
          >
            3
          </span>
        </div>
        <ul
          className="property__features"
        >
          <li
            className="property__feature property__feature--entire"
          >
            Test Type
          </li>
          <li
            className="property__feature property__feature--bedrooms"
          >
            5
             Bedrooms
          </li>
          <li
            className="property__feature property__feature--adults"
          >
            Max 
            5
             adults
          </li>
        </ul>
        <div
          className="property__price"
        >
          <b
            className="property__price-value"
          >
            €
            200
          </b>
          <span
            className="property__price-text"
          >
             night
          </span>
        </div>
        <div
          className="property__inside"
        >
          <h2
            className="property__inside-title"
          >
            What's inside
          </h2>
          <ul
            className="property__inside-list"
          >
            <li
              className="property__inside-item"
              key="0"
            >
              Good1
            </li>
            <li
              className="property__inside-item"
              key="1"
            >
              Good2
            </li>
          </ul>
        </div>
        <div
          className="property__host"
        >
          <h2
            className="property__host-title"
          >
            Meet the host
          </h2>
          <div
            className="property__host-user user"
          >
            <div
              className="property__avatar-wrapper property__avatar-wrapper--pro user__avatar-wrapper"
            >
              <img
                alt="Host avatar"
                className="property__avatar user__avatar"
                height="74"
                src="testAvatar"
                width="74"
              />
            </div>
            <span
              className="property__user-name"
            >
              testName
            </span>
            <span
              className="property__user-status"
            >
              Pro
            </span>
          </div>
          <div
            className="property__description"
          >
            <p
              className="property__text"
            >
              Test Description
            </p>
          </div>
        </div>
        <ReviewsList
          comments={
            Array [
              Object {
                "comment": "comment",
                "date": "date",
                "rating": 4,
                "user": Object {
                  "avatar": "avatar",
                  "isPro": true,
                  "name": "name",
                },
              },
            ]
          }
          id={1}
          isAuthorizationRequired={true}
        />
      </div>
    </div>
    <PlacesMap
      city="Test City"
      className="property"
      places={
        Array [
          Object {
            "bedrooms": 5,
            "coordinates": Array [
              100,
              300,
            ],
            "description": "Test Description",
            "goods": Array [
              "Good1",
              "Good2",
            ],
            "host": Object {
              "avatar": "testAvatar",
              "isPro": true,
              "name": "testName",
            },
            "id": 1,
            "images": Array [
              "Img1",
              "Img2",
            ],
            "img": "test img",
            "isPremium": true,
            "maxAdults": 5,
            "price": 200,
            "rating": 3.1,
            "title": "Test Title",
            "type": "Test Type",
          },
        ]
      }
    />
  </section>
  <div
    className="container"
  >
    <section
      className="near-places places"
    >
      <h2
        className="near-places__title"
      >
        Other places in the neighbourhood
      </h2>
      <WithActiveItem
        classHelper="near"
        places={
          Array [
            Object {
              "bedrooms": 5,
              "coordinates": Array [
                100,
                300,
              ],
              "description": "Test Description",
              "goods": Array [
                "Good1",
                "Good2",
              ],
              "host": Object {
                "avatar": "testAvatar",
                "isPro": true,
                "name": "testName",
              },
              "id": 1,
              "images": Array [
                "Img1",
                "Img2",
              ],
              "img": "test img",
              "isPremium": true,
              "maxAdults": 5,
              "price": 200,
              "rating": 3.1,
              "title": "Test Title",
              "type": "Test Type",
            },
          ]
        }
      />
    </section>
  </div>
</main>
`;
